Geo/11771 fix code smells in dgeflowcpp reported by sonarcloud

indigocoral opened the pull request:
**📝 Description**
Fixed some code smells in 'dgeflow.cpp' and 'dgeflow.h' files

**🆕 Changelog**

- broken down "ExecuteFlowAnalysis" into separate functions to reduce the Cognitive Complexity
- added 'CriticalHeadInfo' structure to reduce the parameters given to the functions
- added some parameters as a member 
- made some of the functions 'const' 
- replaced 'new' by 'make_shared'
- removed a 'catch' as it was not doing anything
- replaced 'std::cerr' with 'KRATOS_ERROR_IF_NOT(RiverBoundary)' since it was in the if condition: "if (!RiverBoundary)"
- added a function for logging output
- fixed code smell regarding number of nested "break" statements
- built 'CallBackFunctions' structure to reduce code smells
- replaced 2 const references to "std::string" by a "std::string_view"
- replaced the type aliasing to "using" instead of "typedef" 
- checked usages of aliases and removed those not being used
- reduced number of parameters given to 'ExecuteFlowAnalysis' function in "dgeflow.cpp" and in "kratos_external_bindings.cpp"
- modified two cpp test files because of the change in parameters given to the 'ExecuteFlowAnalysis' function in "test_head_extrapolation_flow_workflow.cpp" and "test_GeoMechanicsNewtonRaphsonErosionProcessStrategy.cpp" test

indigocoral committed:
1. "ExecuteFlowAnalysis" is broken down to 3 function:
- ExecuteWithoutPiping
- ExecuteWithPiping
- AddNodalSolutionStepVariables
2. Made a first attempt to reduce the parameters given to the functions

indigocoral committed:
Added a 'CriticalHeadInfo' structure to reduce the parameters given to the functions

indigocoral committed:
- added rCriticalHeadBoundaryModelPartName parameter as a member
- moved gid_output_settings out of the piping functions

indigocoral committed:
removed unnecessary parameters from 'ExecuteWithoutPiping' function

indigocoral committed:
- made some of the functions 'const'
- 'new' replaced by 'make_shared'
- removed 'catch' as it was not doing anything

indigocoral committed:
replaced 'std::cerr' with 'KRATOS_ERROR_IF_NOT(RiverBoundary)' since it was in the if condition: "if (!RiverBoundary)"

indigocoral committed:
- replaced 'new' with 'make_shared' for "p_output"
- moved some part of the 'ExecuteWithPiping' function to another function

indigocoral committed:
added 'criticalHead' as a private member

indigocoral committed:
added a function for logging output

indigocoral committed:
- added 'HandleCriticalHeadFound' function
- fixed code smell regarding number of nested "break" statements
- renamed some functions for better understanding
- introduced 'currentHead' & 'exitLoop' as members

indigocoral committed:
fixed the error in the build server

indigocoral committed:
built 'struct CallBackFunctions' to reduce code smells

indigocoral committed:
made 'processes' a member to reduce the parameters given to the functions

indigocoral committed:
replaced 2 const references to "std::string" by a "std::string_view"

indigocoral committed:
- replaced the type aliasing to "using" instead of "typedef"
- checked usages of aliases and removed those not being used
- corrected member variable names

indigocoral committed:
moved structures to the header file

indigocoral committed:
- reduced number of parameters given to 'ExecuteFlowAnalysis' function in "dgeflow.cpp" and in "kratos_external_bindings.cpp"
- moved the structure to the header file
- modified two cpp test files because of the change in parameters given to the 'ExecuteFlowAnalysis' function in "test_head_extrapolation_flow_workflow.cpp" and "test_GeoMechanicsNewtonRaphsonErosionProcessStrategy.cpp" test

indigocoral committed:
Merge remote-tracking branch 'origin/master' into geo/11771-fix-a-code-smell-in-dgeflowcpp-reported-by-sonarcloud

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_GeoMechanicsNewtonRaphsonErosionProcessStrategy.cpp:
```diff
@@ -34,9 +34,14 @@ KRATOS_TEST_CASE_IN_SUITE(ErosionProcessStrategy, KratosGeoMechanicsIntegrationS
+    Kratos::KratosExecute::CallBackFunctions call_back_functions(&flow_stubs::emptyLog,
+                                                                 &flow_stubs::emptyProgress,
+                                                                 &flow_stubs::emptyLog,
+                                                                 &flow_stubs::emptyCancel);
```
These two structs can be const (same for the other ones in this file and in the other test file)
```suggestion
    const Kratos::KratosExecute::CriticalHeadInfo critical_head_info(3, 4, 0.1);
    const Kratos::KratosExecute::CallBackFunctions call_back_functions(&flow_stubs::emptyLog,
                                                                 &flow_stubs::emptyProgress,
                                                                 &flow_stubs::emptyLog,
                                                                 &flow_stubs::emptyCancel);
```

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_GeoMechanicsNewtonRaphsonErosionProcessStrategy.cpp:
```diff
@@ -34,9 +34,14 @@ KRATOS_TEST_CASE_IN_SUITE(ErosionProcessStrategy, KratosGeoMechanicsIntegrationS
+                                                                 &flow_stubs::emptyLog,
+                                                                 &flow_stubs::emptyCancel);
+
+    int status = execute.ExecuteFlowAnalysis(workingDirectory, projectFile, critical_head_info, "PorousDomain.Left_head", call_back_functions);
```
Also this int can be const (same holds for the other tests)
```suggestion
    const int status = execute.ExecuteFlowAnalysis(workingDirectory, projectFile, critical_head_info, "PorousDomain.Left_head", call_back_functions);
```

rfaasse commented on applications/GeoMechanicsApplication/external_bindings/kratos_external_bindings.cpp:
```diff
@@ -24,16 +24,17 @@ extern "C"
+        Kratos::KratosExecute::CallBackFunctions call_back_functions(logCallback,
+                                                                     reportProgress,
+                                                                     reportTextualProgress,
+                                                                     shouldCancel);
```
These can be const
```suggestion
        const Kratos::KratosExecute::CriticalHeadInfo critical_head_info(minCriticalHead, maxCriticalHead, stepCriticalHead);
        const Kratos::KratosExecute::CallBackFunctions call_back_functions(logCallback,
                                                                     reportProgress,
                                                                     reportTextualProgress,
                                                                     shouldCancel);
```

rfaasse commented on applications/GeoMechanicsApplication/external_bindings/kratos_external_bindings.cpp:
```diff
@@ -24,16 +24,17 @@ extern "C"
+                                                                     reportProgress,
+                                                                     reportTextualProgress,
+                                                                     shouldCancel);
+
```
Here we can directly return instance->ExecuteFlowAnalysis( ... ) instead of saving it in errorCode

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
+        std::string mCriticalHeadBoundaryModelPartName;
+        bool mPipingSuccess = false;
+        double mCriticalHead;
+        double mCurrentHead;
```
Here I would give some default value (not sure if these would be sensible?)
```suggestion
        double mCriticalHead = 0.0;
        double mCurrentHead = 0.0;
```

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
         static LinearSolverType::Pointer setup_solver_dgeoflow();
         static GeoMechanicsNewtonRaphsonErosionProcessStrategyType::Pointer setup_strategy_dgeoflow(ModelPart &model_part);
-        static std::vector<std::shared_ptr<Process>> parseProcess(ModelPart &model_part, Parameters projFile);
+        void parseProcess(ModelPart &model_part, Parameters projFile);
```
I think it would be better to rename this function to ParseProcesses (since it initializes multiple of them and to be in line with the style guide)

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
+
         void ResetModelParts();
 
         int echoLevel = 1;
```
Let's move this one to all the other members and call it mEchoLevel

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
         // Initial Setup
         Model current_model;
         Kernel kernel;
         KratosGeoMechanicsApplication::Pointer geoApp;
```
These members do not yet adhere to the style guide
```suggestion
        Model mCurrentModel;
        Kernel mKernel;
        KratosGeoMechanicsApplication::Pointer mpGeoApp;
```

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
 
-                // Close the file
-                CriticalHeadFile.close();
+    void KratosExecute::AddNodalSolutionStepVariables(ModelPart& model_part) const
```
To adhere to the style-guide (which is a bit confusing, I agree), model_part should be renamed to rModelPart

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
             KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Parsed Process Data" << std::endl;
 
-            bool hasPiping = stepCriticalHead != 0;
+            bool hasPiping = criticalHeadInfo.stepCriticalHead != 0;
```
Minor style guide thing: hasPiping -> has_piping

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -254,106 +250,79 @@ namespace Kratos
-                                           const std::function<bool()>&             rShouldCancel)
+    int KratosExecute::ExecuteFlowAnalysis(std::string_view rWorkingDirectory,
+                                           const std::string& rProjectParamsFileName,
+                                           const CriticalHeadInfo& criticalHeadInfo,
```
Minor style guide comment: criticalHeadInfo -> rCriticalHeadInfo

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
-                rLogCallback(kratosLogBuffer.str().c_str());
-                Logger::RemoveOutput(p_output);
-                ResetModelParts();
+                HandleCancellationAndReset(rCallBackFunctions.rLogCallback, p_output);
```
I really like you extracted this HandleCancellationAndReset, which you have re-used in several places!

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
-                {
-                    RiverBoundary = FindRiverBoundaryByName(rCriticalHeadBoundaryModelPartName, processes);
-                }
+            HandleCancellationAndReset(rCallBackFunctions.rLogCallback, p_output);
```
Since here, there is no cancellation, we might need to make the name of the function a bit more general. Maybe HandleCleanUp()?

indigocoral committed:
removed the extra lines added for debugging

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+                                         LoggerOutput::Pointer p_output,
+                                         const CallBackFunctions& rCallBackFunctions)
+    {
+        std::stringstream kratosLogBuffer;
```
kratosLogBuffer can be kratos_log_buffer, since it's a local variable

avdg81 replied:
I would actually remove it, since it is an unused variable in this function.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+        KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head step size: " << criticalHeadInfo.stepCriticalHead << std::endl;
+        const auto p_solving_strategy = setup_strategy_dgeoflow(model_part);
+        p_solving_strategy->SetEchoLevel(0);
+        shared_ptr<Process> RiverBoundary;
```
Can be river_boundary to adhere to the style guide

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
 
-                auto currentProcess = std::static_pointer_cast<GeoFlowApplyConstantHydrostaticPressureProcess>(RiverBoundary);
-                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "River boundary name: " << currentProcess->GetName() << std::endl;
+    int KratosExecute::ExecuteWithPiping(ModelPart& model_part,
```
We can change this to rModelPart to adhere to the style guide

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
 
-                int step = 1;
-                const auto maxSteps = static_cast<int>(std::ceil((maxCriticalHead - minCriticalHead) / stepCriticalHead)) + 2;
+        KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Writing result to: " << mWorkingDirectory << "\\criticalHead.json" << std::endl;
```
I would extract this part (423-443) with the CriticalHeadFile into a function, since it is a very clear singular thing that's going on, which makes sense to group in a single function.

rfaasse replied:
Nicely done, thanks!

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
-                    }
-                }
+        // output critical head_json
+        std::ofstream CriticalHeadFile(mWorkingDirectory + "\\criticalHead.json");
```
To adhere to the style guide, this should be critical_head_file

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
 
-                // output critical head_json
-                std::ofstream CriticalHeadFile(rWorkingDirectory + "\\criticalHead.json");
+        // Close the file
```
This comment is not really necessary, since the next line CriticalHeadFile.close() describes very clearly what is going on

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+    {
 
-                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Finished writing result" << std::endl;
+        auto currentProcess = std::static_pointer_cast<GeoFlowApplyConstantHydrostaticPressureProcess>(RiverBoundary);
```
to adhere to the style guide, this should be current_process

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+                                        const Kratos::Parameters& gid_output_settings,
+                                        const CriticalHeadInfo& criticalHeadInfo,
+                                        LoggerOutput::Pointer p_output,
+                                        const shared_ptr<Process>& RiverBoundary,
```
Since the river boundary is already a shared_ptr, in my opinion it is not necessary to make it a '&' as well. Do you agree @avdg81?

That means I think it should be const std::shared_ptr<Process> pRiverBoundary (also note the p for pointer)

avdg81 replied:
Yes, I agree. We could even consider to make it a raw pointer or a reference, since we don't intend to share ownership here. But perhaps it's better to leave that for another PR. The focus here is removing code smells.

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+                                        const CriticalHeadInfo& criticalHeadInfo,
+                                        LoggerOutput::Pointer p_output,
+                                        const shared_ptr<Process>& RiverBoundary,
+                                        const GeoMechanicsNewtonRaphsonErosionProcessStrategyType::Pointer& p_solving_strategy,
```
Same here, I don't think the '&' is necessary. For the style guide, this should be pSolvingStrategy

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+        mCriticalHead = mCurrentHead;
+
+        std::vector<Element *> pipeElements;
+        pipeElements = p_solving_strategy->GetPipingElements();
```
This could be one line (also note the renaming to pipe_elements): 
```suggestion
        auto pipe_elements = p_solving_strategy->GetPipingElements();
```

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+        const auto noPipeElements = pipeElements.size();
+
+        int step = 1;
+        const auto maxSteps = static_cast<int>(std::ceil((criticalHeadInfo.maxCriticalHead - criticalHeadInfo.minCriticalHead) / criticalHeadInfo.stepCriticalHead)) + 2;
```
This can be max_steps

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+        const auto maxSteps = static_cast<int>(std::ceil((criticalHeadInfo.maxCriticalHead - criticalHeadInfo.minCriticalHead) / criticalHeadInfo.stepCriticalHead)) + 2;
+
+
+        while (!mExitLoop)
```
Is this mExitLoop used outside of this function? I think for this one might be better to keep it a local variable within this function. Maybe we can also discuss if it's necessary (was one of the SQ issues the 'while(true' statement?

rfaasse replied:
Solved in a different way now, nicely done!

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+    int KratosExecute::FindCriticalHead(ModelPart& model_part,
+                                        const Kratos::Parameters& gid_output_settings,
+                                        const CriticalHeadInfo& criticalHeadInfo,
+                                        LoggerOutput::Pointer p_output,
```
Also here it can be rModelPart, rGidOutputSettings, rCriticalHeadInfo and pOutput

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
 
-            ResetModelParts();
-            return 0;
+            std::ostringstream currentHeadStream;
```
This could be current_head_stream

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
-            return 0;
+            std::ostringstream currentHeadStream;
+            currentHeadStream << std::setprecision(8) << std::noshowpoint << mCurrentHead;
+            std::string currentHeadString = currentHeadStream.str();
```
This can be current_head_string

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+        return 0;
+    }
+
+    void KratosExecute::HandleCriticalHeadFound(const CriticalHeadInfo& criticalHeadInfo)
```
Can be rCriticalHeadInfo

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -566,15 +608,14 @@ namespace Kratos
-            std::cerr << "No boundary found with the model part name " << rCriticalHeadBoundaryModelPartName << "." << std::endl;
+            KRATOS_ERROR_IF_NOT(RiverBoundary) << "No boundary found with the model part name " << rCriticalHeadBoundaryModelPartName << "." << std::endl;
             return nullptr;
         }
```
This can be simplified, since KRATOS_ERROR_IF_NOT throws, so this could be changed to just 
KRATOS_ERROR_IF_NOT(RiverBoundary) << "No boundary found with the model part name " << rCriticalHeadBoundaryModelPartName << "." << std::endl;

And remove the if statement and the return nullptr

indigocoral committed:
attempt to fix the GitHub failing error

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -623,21 +657,15 @@ namespace Kratos
-            std::cerr << "No boundary found on the river side at node " << RiverNode << "." << std::endl;
+            KRATOS_ERROR_IF_NOT(RiverBoundary) << "No boundary found on the river side at node " << RiverNode << "." << std::endl;
             return nullptr;
         }
```
Same as earlier comment

This can be simplified, since KRATOS_ERROR_IF_NOT throws, so this could be changed to just
KRATOS_ERROR_IF_NOT(RiverBoundary) << "No boundary found with the model part name " << rCriticalHeadBoundaryModelPartName << "." << std::endl;

And remove the if statement and the return nullptr

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -216,33 +214,31 @@ namespace Kratos
         ModelPart &part = model_part.GetSubModelPart(subname);
-        processes.push_back(make_shared<ApplyConstantScalarValueProcess>(ApplyConstantScalarValueProcess(part, VOLUME_ACCELERATION_X,
+        mProcesses.push_back(make_shared<ApplyConstantScalarValueProcess>(ApplyConstantScalarValueProcess(part, VOLUME_ACCELERATION_X,
                                                                                                          0.0, 0, ApplyConstantScalarValueProcess::VARIABLE_IS_FIXED)));
```
Since `make_shared` receives a list of arguments that it will pass on to the constructor of class `ApplyConstantScalarValueProcess`, there is no need to create a temporary object of this type. In other words, you could simplify this as follows:
```suggestion
        mProcesses.push_back(make_shared<ApplyConstantScalarValueProcess>(part, VOLUME_ACCELERATION_X,
                                                                                                         0.0, 0, ApplyConstantScalarValueProcess::VARIABLE_IS_FIXED));
```

rfaasse commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -597,24 +638,17 @@ namespace Kratos
-                    catch (...)
-                    {
-                    }
+                    currentModelPart->GetNode(RiverNode);
```
Here we miss the initialization of currentModelPart, so it will be a nullptr when calling it. We still need the `currentModelPart = &current_process->GetModelPart();` which was there before

rfaasse replied:
By the way, @indigocoral did this not result in one or more failing tests? Then we might not have a test related to this specific part

indigocoral replied:
No, none of the tests failed because of this.

rfaasse reviewed (commented):
Very well done @indigocoral, this cleaned up the code a lot and made it more readable! 

Don't be discouraged by the number of comments, since only one of them is really important (it has to do with an un-initialized pointer of a model part). 

Most of the others are small style/readability improvements and probably 9 out of 10 are on code that was already there, but are easy fixes to make now we're working on it anyway.

All in all, a big step forward for this class!

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
+            std::function<void(const char*)> rLogCallback;
+            std::function<void(double)> rReportProgress;
+            std::function<void(const char*)> rReportTextualProgress;
+            std::function<bool()> rShouldCancel;
```
Since these are no longer references, we should remove the `r` prefix (yes, this is a very annoying result of the needlessly complex style guide).
```suggestion
            std::function<void(const char*)> LogCallback;
            std::function<void(double)> ReportProgress;
            std::function<void(const char*)> ReportTextualProgress;
            std::function<bool()> ShouldCancel;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
+            CallBackFunctions(std::function<void(const char*)> rLogCallback,
+                              std::function<void(double)> rReportProgress,
+                              std::function<void(const char*)> rReportTextualProgress,
+                              std::function<bool()> rShouldCancel) :
```
And the same thing here:
```suggestion
            CallBackFunctions(std::function<void(const char*)> LogCallback,
                              std::function<void(double)> ReportProgress,
                              std::function<void(const char*)> ReportTextualProgress,
                              std::function<bool()> ShouldCancel) :
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
+            {}
+        };
+
+        int ExecuteFlowAnalysis(std::string_view rWorkingDirectory,
```
Again a minor style thing (the parameter is now passed by value and no longer by reference):
```suggestion
        int ExecuteFlowAnalysis(std::string_view WorkingDirectory,
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.h:
```diff
@@ -46,56 +46,109 @@
+        int ExecuteFlowAnalysis(std::string_view rWorkingDirectory,
+                                const std::string& rProjectParamsFileName,
+                                const CriticalHeadInfo& criticalHeadInfo,
+                                std::string_view rCriticalHeadBoundaryModelPartName,
```
And the same thing here:
```suggestion
                                std::string_view CriticalHeadBoundaryModelPartName,
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -216,33 +214,31 @@ namespace Kratos
 
-        processes.push_back(make_shared<ApplyConstantScalarValueProcess>(ApplyConstantScalarValueProcess(part, VOLUME_ACCELERATION_Y, -9.81,
+        mProcesses.push_back(make_shared<ApplyConstantScalarValueProcess>(ApplyConstantScalarValueProcess(part, VOLUME_ACCELERATION_Y, -9.81,
                                                                                                          0, ApplyConstantScalarValueProcess::VARIABLE_IS_FIXED)));
```
And the same simplification here:
```suggestion
        mProcesses.push_back(make_shared<ApplyConstantScalarValueProcess>(part, VOLUME_ACCELERATION_Y, -9.81,
                                                                                                         0, ApplyConstantScalarValueProcess::VARIABLE_IS_FIXED));
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -216,33 +214,31 @@ namespace Kratos
 
-        processes.push_back(make_shared<Process>(ApplyConstantScalarValueProcess(part, VOLUME_ACCELERATION_Z, 0.0,
+        mProcesses.push_back(make_shared<Process>(ApplyConstantScalarValueProcess(part, VOLUME_ACCELERATION_Z, 0.0,
                                                                                  0, ApplyConstantScalarValueProcess::VARIABLE_IS_FIXED)));
```
And also here (note the template parameter now must be updated too):
```suggestion
        mProcesses.push_back(make_shared<ApplyConstantScalarValueProcess>(part, VOLUME_ACCELERATION_Z, 0.0,
                                                                                 0, ApplyConstantScalarValueProcess::VARIABLE_IS_FIXED));
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
-                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head max head: " << maxCriticalHead << std::endl;
-                KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head step size: " << stepCriticalHead << std::endl;
+                ExecuteWithPiping(model_part, gid_output_settings, criticalHeadInfo, p_output, rCallBackFunctions);
+            }
```
Now that you have simplified the bodies of the `if` and `else` clauses, I would invert the logic (because negation is slightly harder to understand than not negating something). I.e. (pseudo-code, to show what I mean):
```
if (hasPiping)
{
    ExecuteWithPiping(...);
}
else
{
    ExecuteWithoutPiping(...);
}
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+
+    void KratosExecute::HandleCriticalHeadFound(const CriticalHeadInfo& criticalHeadInfo)
+    {
+        if (abs(mCurrentHead - criticalHeadInfo.minCriticalHead) < 1e-9)
```
I always prefer to use the math functions from the `std` namespace (to ensure that we don't accidentally call the C-style variants, which don't support overloading). So:
```suggestion
        if (std::abs(mCurrentHead - criticalHeadInfo.minCriticalHead) < 1e-9)
```

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+                                             const Kratos::Parameters& gid_output_settings) const
+    {
+        const auto p_solving_strategy = setup_strategy_dgeoflow(model_part);
+        p_solving_strategy->SetEchoLevel(0);
```
I believe these lines have been accidentally duplicated. They are also present in their original location (lines 312-313), but the strategy created there is now no longer used. I would suggest to pass it to this function through the parameter list (i.e. add an additional function parameter), and to not create it here.

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+        KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head max head: " << criticalHeadInfo.maxCriticalHead << std::endl;
+        KRATOS_INFO_IF("GeoFlowKernel", this->GetEchoLevel() > 0) << "Critical head step size: " << criticalHeadInfo.stepCriticalHead << std::endl;
+        const auto p_solving_strategy = setup_strategy_dgeoflow(model_part);
+        p_solving_strategy->SetEchoLevel(0);
```
As for the member function `ExecuteWithoutPiping`, I believe these lines have been accidentally duplicated. I would suggest to pass the originally created strategy through the parameter list, rather than creating a new one here.

avdg81 commented on applications/GeoMechanicsApplication/custom_workflows/dgeoflow.cpp:
```diff
@@ -371,187 +340,260 @@ namespace Kratos
+            if (count == noPipeElements)
+            {
+                HandleCriticalHeadFound(criticalHeadInfo);
+                mExitLoop = true;
```
There used to be a `break` here, which means that we would jump out of the loop. Now we continue until the end of loop body. I'm not sure whether that was your intention, because the behavior is different now.

indigocoral committed:
applied some review comments

indigocoral committed:
applied some more review comments

indigocoral committed:
applied some more review comments

indigocoral committed:
Merge remote-tracking branch 'origin/master' into geo/11771-fix-a-code-smell-in-dgeflowcpp-reported-by-sonarcloud

indigocoral committed:
applied more review comments

indigocoral committed:
applied minor changes

indigocoral committed:
fixed the GitHub error

avdg81 committed:
Extracted a loop condition for finding the critical head

One of the loop conditions is whether the current head does not exceed
the maximum critical head yet. That logic has been moved to a separate
member function, which makes it easier to understand how the loop works.

As a result, data member `mExitLoop` was removed, since it had become
redundant.

indigocoral committed:
replaced 'mCriticalHead' with 'mCurrentHead' in the "AreExceedingMaxCriticalHead" function

rfaasse reviewed (approved):
Thanks for making all the changes and cleaning up the code in general! It was a lot of work, but in the end it improved the code a lot!
